src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.h

```diff
@@ -0,0 +1,942 @@
+     * @note For non-list types:
+     *       - Creates default-initialized value
+     *       - Ignores size parameter
+     */
+    CHIP_ERROR CreateNewValue(size_t size)
+    {
+        if (mUpdateState != UpdateState::kIdle)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+
+        if constexpr (IsValueList() || IsList<WrappedType>::value)
+        {
+            if (size >= 1)
+            {
+                auto * buffer = static_cast<PayloadType *>(chip::Platform::MemoryCalloc(size, sizeof(PayloadType)));
+
+                if (!buffer)
+                {
+                    return CHIP_ERROR_NO_MEMORY;
+                }
+
+                if constexpr (IsValueNullable())
+                {
+                    GetNewValueRef().SetNonNull(DataModel::List<PayloadType>(buffer, size));
+                }
+                else
+                {
+                    GetNewValueRef() = DataModel::List<PayloadType>(buffer, size);
+                }
+            }
+            else
+            {
+                if constexpr (IsValueNullable())
+                {
+                    GetNewValueRef().SetNull();
+                }
+            }
+        }
+        else if constexpr (IsValueNullable())
+        {
+            WrappedType tmp = WrappedType();
+            GetNewValueRef().SetNonNull(tmp);
+        }
+        else
+        {
+            GetNewValueRef() = ValueType();
+        }
+
+        mUpdateState = UpdateState::kInitialized;
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Accesses the pending new value's payload for modification
+     * @return Pointer to modifiable payload value if available
+     * @retval nullptr if:
+     *         - Not in kInitialized state
+     *         - Value is null (for nullable types)
+     *         - List is empty (for list types)
+     *
+     * @warning Pointer becomes invalid after UpdateCommit()/UpdateEnd()
+     */
+    auto GetNewValueData() -> ExtractPayloadType_t<ValueType> *
+    {
+        if (mUpdateState < UpdateState::kInitialized)
+        {
+            return nullptr;
+        }
+
+        if constexpr (IsValueNullable())
+        {
+            if (GetNewValueRef().IsNull())
+            {
+                return nullptr;
+            }
+            if constexpr (IsList<WrappedType>::value)
+            {
+                return GetNewValueRef().Value().data();
+            }
+            else
+            {
+                return &GetNewValueRef().Value();
+            }
+        }
+        else if constexpr (IsValueList())
+        {
+            return GetNewValueRef().data();
+        }
+        else
+        {
+            return &GetNewValueRef();
+        }
+    }
+
+    /**
+     * @brief Signals completion of new value modifications
+     * @return CHIP_NO_ERROR on success
+     * @retval CHIP_ERROR_INCORRECT_STATE if not in kInitialized state
+     */
+    CHIP_ERROR MarkAsAssigned()
```

Comment:
I'm still rather confused about this state machine that is not actually followed by TariffDataUpdate.

Reply:
https://github.com/project-chip/connectedhomeip/issues/40119